File: script/process.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat
from scipy.interpolate import interp1d

# offset
file = loadmat("empty.mat")
data = file["data"]
time = file["timestamp"]
acc = data[:, 0:3]
force = data[:, 3:5]
offset_force = np.mean(force, axis=0)

plt.figure(3)
plt.plot(time, force - offset_force)
plt.show()


file = loadmat("pos2_flex.mat")
data = file["data"]
time = file["timestamp"]
acc = data[:, 0:3]
force = data[:, 3:5]
offset_force = np.mean(force, axis=0)

plt.figure(2)
plt.plot(time, acc[:,0])
plt.show()


types = ["relax", "flex"]
conf = [ "2", "3"]
angles = np.array([85, 110]) + 90

acceleration = []

def zci(v):
    return np.where(np.diff(np.sign(v)))[0]

for i in range(len(conf)):
    for j in range(len(types)):
        filename = "pos" + conf[i] + "_" + types[j]
        file = loadmat(filename)
        data = np.column_stack((file["timestamp"], file["data"]))
        data = data[np.argsort(data[:, 0])]
        time = data[:, 0].astype(float) / 1e6
        acc = data[:, 1:4].astype(float)
        force = data[:, 4:6] - offset_force
        
        x = np.arange(time[0], time[-1], 1 / 1000)
        per = 0
        acceleration = []
        alpha = (angles[i] - 40) * np.pi / 180
        for k in range(2):
            y = interp1d(time, acc[:, k], kind="linear")(x)
    
            ym = np.mean(y)
            yz = y - ym
            zx = x[zci(yz)]
            if per == 0:
                per = 2 * np.mean(np.diff(zx))
            
            peak = zci(yz) + int(per * 1000 / 4)
            A = np.mean(np.abs(yz[peak[:-10]]))
            
            shift = 0
            yp = A * np.sin(2 * np.pi / per * x) + ym
            min_sum = np.sum(np.abs(yp - y))
            for s in np.arange(0, per, 0.001):
                yp = A * np.sin(2 * np.pi / per * (x + s)) + ym
                su = np.sum(np.abs(yp - y))
                if su < min_sum:
                    min_sum = su
                    shift = s
            yp = A * np.sin(2 * np.pi / per * (x + shift)) + ym

            acceleration.append(yp)
        
        acceleration = np.array(acceleration)
        x = acceleration[0] * np.cos(alpha) + acceleration[1] * np.sin(alpha) + 500 * np.cos(np.pi * angles[i] / 180)
        y = -acceleration[1] * np.cos(alpha) + acceleration[0] * np.sin(alpha) + 500 * np.sin(np.pi * angles[i] / 180)
        
        co = "b"
        if j == 1:
            co = "r"
        
        plt.figure(1)
        plt.plot(x, y, co)

        plt.figure(4+j)
        plt.subplot(3, 1, i + 1)
        me = np.mean(force, axis=1)
        print(me.shape, time.shape)
        plt.plot(time[9:], me[9:], co)
        plt.title(f"Force arm angle: {angles[i]}deg")
        plt.legend(["Relaxing", "Flexing"])
        
plt.figure(1)
for i in range(len(conf)):
    plt.plot([0, 500 * np.cos(np.pi * angles[i] / 180)], [0, 500 * np.sin(np.pi * angles[i] / 180)], "-o")

plt.plot([0, 0], [-500, 0], "black")
plt.title("Acceleration for 3 arm configurations")
plt.legend(["Relaxing", "Flexing"])

m = 700
#set the limits of the plot to plus or minus m
plt.xlim(-m, m)
plt.ylim(-m, m)
plt.show()

File: script/process.m
%% offset
file = load("empty.mat");
data = file.data;
time = file.timestamp;
acc = data(:, 1:3);
force = data(:, 4:5);
offset_force = mean(force);
plot(time, force-offset_force)
%%
types = ["relax","flex"];
conf=["1", "2", "3"];
angles=[0,85,110]+90;

acceleration = [];
zci = @(v) find(v(:).*circshift(v(:), [-1 0]) <= 0);  

% w=2*pi/per;
% C=m*r*w^2
% C/Acos(phi) = D;
% C/Asin(phi) = (Iw - K/w);

for i = 1:length(conf)
    for j = 1:length(types)
        filename = "pos" + conf(i)+"_"+types(j);
        file = load(filename);
        data = [file.timestamp file.data ];
        data = sortrows(data);
        time = double(data(:,1))/1000000;
        acc = double(data(:, 2:4));
        force = data(:, 5:6)-int64(offset_force);

        
        
        x = time(1):1/1000:time(end);
        per=0;
        acceleration = [];
        alpha=(angles(i)-40)*pi/180;
        for k = 1:2
            y = interp1(time,acc(:,k),x);
    
            ym = mean(y);                    
            yz = y-mean(y);
            zx = x(zci(yz));     % Find zero-crossing
            if per==0
            per = 2*mean(diff(zx));                     % Estimate period
            end
           
            peak=zci(yz)+floor(per*1000/4);
            A=mean(abs(yz(peak(1:end-10))));                         
            
            shift = 0;
            yp = A*sin(2*pi/per*(x))+ym;
            min_sum=sum(abs(yp-y));
            for s = 0:0.001:per
                yp = A*sin(2*pi/per*(x+s))+ym;
                su=sum(abs(yp-y));
                if su < min_sum
                    min_sum = su;
                    shift=s;
                end
            end
            yp = A*sin(2*pi/per*(x+shift))+ym;


            if(length(acceleration)==0)
                acceleration=yp;
            else
                acceleration=[acceleration; yp];
            end
        end
        x=acceleration(1,:).*cos(alpha)+ acceleration(2,:).*sin(alpha)+500*cos(pi*angles(i)/180);
        y=-acceleration(2,:).*cos(alpha)+ acceleration(1,:).*sin(alpha)+500*sin(pi*angles(i)/180);
        %figure(1)
        co="b"
        if j==2
            co="r"
        end
        figure(1)
        plot(x, y, co)
        hold on;

        figure(2)
        subplot(3,1,i)
        me = mean(force,2)
        plot(time(10:end),me(10:end), co)
        title("Force arm angle: " + angles(i) + "deg")
        legend("Relaxing", "Flexing")
        hold on
        
    end
end

figure(1)
for i = 1:length(conf)
    plot([0 500*cos(pi*angles(i)/180)], [0 500*sin(pi*angles(i)/180)], "-o")
end
figure(1)
plot([0 0], [-500 0], "-oblack")
title("Acceleration for 3 arm configurations")
legend("Relaxing", "Flexing")


m=700
figure(1)
axis([-m m -m m], "equal")
hold off;                 % Returns Zero-Crossing Indices Of Argument Vector

